text/0002-time-based-resolution-mode.md

```diff
@@ -0,0 +1,48 @@
+# Time-based resolution mode
+
+## Summary
+
+A new resolution mode that resolves versions not to their highest versions but to versions that were published in a given time frame.
+
+## Motivation
+
+The resolution mode that is currently used by pnpm is not optimized for caching. During installation we need to always fetch the metadata of each package as we don't know if the metadata in the cache has the highest version of the package. Also, installation of the highest version is dangerous as someone may hijack a subdependency, publish a new version, and pnpm will immediately install it.
+
+## Detailed Explanation
+
+Resolution will be divided into two stages. The first stage will resolve all the direct dependencies of all the workspace projects. This stage may work the same way as it works now (highest version is picked). When all the direct dependencies are resolved, we check when were the picked versions released. This information is present in the package metadata at the "time" field. For example, if we install webpack and eslint, we get webpack resolved to v5.74.0 and eslint resolved to v8.22.0. `webpack@5.74.0` was released at "2022-07-25T08:00:33.823Z". `eslint@8.22.0` was released at "2022-08-14T01:23:41.730Z". Now we compare the dates of each released packages and pick the nearest date. In this case, the nearest date is the date eslint was released: "2022-08-14T01:23:41.730Z". We add some delta (e.g. 1 hour) to this time and call the resulting date T. The delta is needed because in some monorepos packages may be published in random order, so dependent packages may be released after dependency packages.
+
+At the second stage, we resolve all the subdependencies. At this stage, instead of resolving a range to the highest available version, we filter out any versions that were released after T and pick the highest version from those. Let's say we need to resolve `ajv@^6.10.0`. We already have a metadata of ajv in cache and it was saved after T, so we don't need to redownload it. This are the versions of ajv that match `^6.10.0`:
+
+| version | release date |
+|--|--|
+| 6.10.0 | 2021-01-08 |
+| 6.10.1 | 2021-12-22 |
+| 6.10.2 | 2022-01-25 |
+| 6.11.0 | 2022-04-05 |
+| 6.11.1 | 2022-08-16 |
+
+As you can see, v6.11.1 was released after T, so we pick v6.11.0, which is the highest version that satisfies the range and the time frame.
```

Comment:
Could you elaborate why`so we pick v6.11.0` was picked up? Above says `^6.10.0`, how does `v6.11.0` get picked, I'n not sure if I missing something.

Reply:
Nevermind, my bad,  I was thinking on `~6.10.0`.

Reply:
https://github.com/pnpm/pnpm/pull/5298